ledger/src/blockstore.rs

```diff
@@ -3315,6 +3315,54 @@ impl Blockstore {
         );
         Ok(())
     }
+
+    /// Mark a root `slot` as connected, traverse `slot`'s children and update
+    /// the children's connected status if appropriate.
+    ///
+    /// A ledger with a full path of blocks from genesis to the latest root will
+    /// have all of the rooted blocks marked as connected such that new blocks
+    /// could also be connected. However, starting from some root (such as from
+    /// a snapshot) is a valid way to join a cluster. For this case, mark this
+    /// root as connected such that the node that joined midway through can
+    /// have their slots considered connected.
+    pub fn set_and_chain_connected_on_root_and_next_slots(&self, root: Slot) -> Result<()> {
```

Comment:
Somehow I feel this should be implemented using two functions: set_connected_on_root() and chain_next_slots()?  The former set both connected and parent_connected to true, and the latter propagates the connectivity.  Since I feel once we've correctly set connected() and parent_connected() to true for the snapshot slot, then the propagation should be the same as the usual case.  Am I right on this part?

Reply:
> Somehow I feel this should be implemented using two functions: set_connected_on_root() and chain_next_slots()?
Well, we essentially do this but not in two separate functions for reasons I'll go into

> The former set both connected and parent_connected to true
This function currently does this, but not with a dedicated `SlotMeta` function. If you feel strongly enough about it, we can add it

> and the latter propagates the connectivity. Since I feel once we've correctly set connected() and parent_connected() to true for the snapshot slot, then the propagation should be the same as the usual case. Am I right on this part?

Yes, the propagation is essentially the same. However, we can't cleanly re-use the existing traversal function because that function will pull `Meta`'s from either the working set (that is built up during insertion) or from the db. This function is only pulling things from the DB, and then committing to a `WriteBatch`.

That being said, these functions utilize the same function that updates a meta and then says whether we should traverse its' children. The duplicated part is just the simple BFS of popping/pushing to a queue.